# __Fire Causes__
#### _Bureau of Fire Investigations_
<div>
<img src="https://upload.wikimedia.org/wikipedia/commons/2/25/Emblem_of_the_New_York_City_Fire_Department.svg" width="120"/>
</div>

<hr>

Source: _<a href="https://data.cityofnewyork.us/Public-Safety/Bureau-of-Fire-Investigations-Fire-Causes/ii3r-svjz/about_data">NYC Open Data</a>_ <br>
Last updated: _9 May 2024_ <br>
Accessed: _3 June 2024_

<hr>

## __Imports__

In [96]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

In [97]:
import sys
import os

sys.path.append(r'C:\Users\vanli\# Python Tutorials\10. Practice\helper_functions')

from helper_functions import dataset_check, uvalues_check, text_check

<hr>

## __Getting to know the dataset__

In [98]:
# Set the path to the file
path = 'datasets/001. Bureau of Fire Investigations - Fire Causes/Bureau_of_Fire_Investigations_-_Fire_Causes_20240603.csv'

In [99]:
df = pd.read_csv(path)
df.head()

Case_Year  Case Number       Incident_DateTime        Borough Battalion  \
0       2016        40001  01/01/2016 12:08:00 AM         Queens        50   
1       2016        50001  01/01/2016 12:13:00 AM  Staten Island        21   
2       2016        40002  01/01/2016 12:21:00 AM         Queens        37   
3       2016        60001  01/01/2016 12:29:00 AM          Bronx        15   
4       2016        40003  01/01/2016 12:32:00 AM         Queens        50   

  Community_District Precinct     Incident_Classification  \
0                  8      107    MD Residential Structure   
1                  2      122    PD Residential Structure   
2                  5      104    MD Residential Structure   
3                 12       47  Burn Notification (Albany)   
4                 12      103    PD Residential Structure   

                             Cause_Fire_Description Fire_Code_Category  
0                         Smoking (Cigarette/Cigar)            Smoking  
1                 Incendiary - Combustible Material         Incendiary  
2                 Incendiary - Combustible Material         Incendiary  
3  Albany Burn Notif. Unrelated to a Fire/Explosion              Other  
4                 Incendiary - Combustible Material         Incendiary

In [100]:
# Set the new column names
colnames = ['Year', 'ID', 'Date', 'Borough', 'Fire battalion', 'Community district', 
            'Police precinct', 'Incident category', 'Fire cause', 'Fire cause category']

In [101]:
# Set the new column names
df = pd.read_csv(path, names = colnames, header = 0)
df.head()

Year     ID                    Date        Borough Fire battalion  \
0  2016  40001  01/01/2016 12:08:00 AM         Queens             50   
1  2016  50001  01/01/2016 12:13:00 AM  Staten Island             21   
2  2016  40002  01/01/2016 12:21:00 AM         Queens             37   
3  2016  60001  01/01/2016 12:29:00 AM          Bronx             15   
4  2016  40003  01/01/2016 12:32:00 AM         Queens             50   

  Community district Police precinct           Incident category  \
0                  8             107    MD Residential Structure   
1                  2             122    PD Residential Structure   
2                  5             104    MD Residential Structure   
3                 12              47  Burn Notification (Albany)   
4                 12             103    PD Residential Structure   

                                         Fire cause Fire cause category  
0                         Smoking (Cigarette/Cigar)             Smoking  
1                 Incendiary - Combustible Material          Incendiary  
2                 Incendiary - Combustible Material          Incendiary  
3  Albany Burn Notif. Unrelated to a Fire/Explosion               Other  
4                 Incendiary - Combustible Material          Incendiary

In [102]:
# Perform the basic check
dataset_check(
    df, 
    shape_check = True,
    dup_check = True,
    dtypes_check = True,
    missing_values_check = True,
    unique_values_check = True,
    stats_check = False,
    corr_check = False
)

COLUMNS:
10

ROWS:
53 741

DUPLICATE ROWS:
0

DATATYPES:
Year                    int64
ID                      int64
Date                   object
Borough                object
Fire battalion         object
Community district     object
Police precinct        object
Incident category      object
Fire cause             object
Fire cause category    object
dtype: object

MISSING VALUES:
Borough                 260
Fire battalion         7658
Community district      338
Police precinct         344
Incident category       827
Fire cause               73
Fire cause category      61
dtype: int64

MISSING VALUES IN %:
Borough                 0.48%
Fire battalion         14.25%
Community district      0.63%
Police precinct         0.64%
Incident category       1.54%
Fire cause              0.14%
Fire cause category     0.11%
dtype: object

UNIQUE VALUES:
Year                       9
ID                      9543
Date                   52343
Borough                    6
Fire battalion           

## __Early findings__

Let’s check what we have:
- <b>Duplicates.</b> There are no duplicate rows.
- <b>Datatypes.</b> Columns <code>Date</code>, <code>Fire battalion</code>, <code>Community district</code>, and <code>Police precinct</code> are seen as <code>object</code> columns even though they contain dates or numbers. This should be investigated.
- <b>Missing values.</b> We have relatively many missing values in column <code>Incident category</code> and we have _a lot_ of missing values in <code>Fire battalion</code>. This is very strange — how come _so many_ incidents do not have an assigned battalion. Does this mean that some fires in the city were not reported/were not handled by the Fire Department? 
- <b>Unique values.</b> We have 53 741 rows, but the <code>ID</code> column has 9 543 unique values. That means that the column is of no use and can be dropped. Also, New York has 5 boroughs (see image below) and not 6. Also a point of interest.

<div>
<img src="https://www.loumovesyou.com/wp-content/uploads/2022/11/FiveBoroughs-01.jpg" width="500"/>
</div>

## __Unique values per column__

Let’s check what are the unique values per column.

In [103]:
def uvalues_check(dataset, columns):
    print('UNIQUE VALUES\n')
    for column in columns:
        print(column.upper())
        uvalues = dataset[column].unique()
        uvalues_sorted = sorted(uvalues, key = lambda x: str(x).lower())
        print(uvalues_sorted, '\n')

In [104]:
df.columns

Index(['Year', 'ID', 'Date', 'Borough', 'Fire battalion', 'Community district',
       'Police precinct', 'Incident category', 'Fire cause',
       'Fire cause category'],
      dtype='object')

In [105]:
uvalues_check(df, ['Year', 'Borough', 'Fire battalion', 'Community district',
                   'Police precinct', 'Incident category', 'Fire cause',
                   'Fire cause category'])

UNIQUE VALUES

YEAR
[2016, 2017, 2018, 2019, 2020, 2021, 2022, 2023, 2024] 

BOROUGH
['Bronx', 'Brooklyn', 'Manhattan', nan, 'Outside of NYC', 'Queens', 'Staten Island'] 

FIRE BATTALION
['0', '1', '10', '11', '12', '13', '14', '15', '16', '17', '18', '19', '2', '20', '21', '22', '23', '26', '27', '28', '3', '31', '32', '33', '35', '37', '38', '39', '4', '40', '41', '42', '43', '44', '45', '46', '47', '48', '49', '50', '51', '52', '53', '54', '55', '57', '58', '6', '7', '8', '9', 'BC', 'CD', 'MR', nan, 'OP', 'SC', 'TR'] 

COMMUNITY DISTRICT
['1', '10', '11', '12', '13', '14', '15', '16', '17', '18', '2', '3', '4', '5', '6', '7', '8', '9', nan, 'Out of NYC'] 

POLICE PRECINCT
['0', '1', '10', '100', '101', '102', '103', '104', '105', '106', '107', '108', '109', '11', '110', '111', '112', '113', '114', '115', '12', '120', '121', '122', '123', '13', '14', '17', '18', '19', '2', '20', '21', '22', '23', '24', '25', '26', '28', '30', '32', '33', '34', '40', '41', '42', '43', '44', '45', '46'

Let’s also check columns <code>Fire battalion</code>, <code>Community district</code>, and <code>Police precinct</code> for textual values.

## __A deeper look at (potentially) numerical columns__

In [106]:
def text_check(dataset, columns):
    df = dataset
    print('NON-NUMERIC VALUES\n')
    for column in columns:
        numeric = pd.to_numeric(df[column], errors = 'coerce')
        text_values = df[column][numeric.isna() & df[column].notna()].unique()
        text_values_sorted = sorted(text_values, key = lambda x: x.lower())
        
        if len(text_values_sorted) > 0:
            print(column.upper())
            print(text_values_sorted, '\n')
        else:
            print(column.upper())
            print('Numeric values only\n')

In [107]:
text_check(df, ['Fire battalion', 'Community district', 'Police precinct'])

NON-NUMERIC VALUES

FIRE BATTALION
['BC', 'CD', 'MR', 'OP', 'SC', 'TR'] 

COMMUNITY DISTRICT
['Out of NYC'] 

POLICE PRECINCT
['5th', '7th', 'HWY2', 'HY2', 'mtn', 'MTS', 'NC05', 'NC08', 'NJ', 'PAPD', 'PSA2', 'PSA7', 'SC01', 'SC07'] 



The <code>ID</code> column can be surely dropped, but what about these textual values we’ve just found out? And what about <code>Outside of NYC</code> in column <code>Borough</code>? Let’s create some pivots to see if these values can be dropped. We can begin with the <code>Borough</code>.

In [108]:
nan_values = 'Unknown'
df['Borough'].fillna(nan_values, inplace = True)

pvt_borough = pd.pivot_table(
    df,
    values = 'ID',
    index = 'Borough',
    aggfunc = 'count',
)

pvt_borough['Percentage'] = (pvt_borough['ID'] / pvt_borough['ID'].sum()) * 100
pvt_borough = pvt_borough.sort_values(by = ['Percentage'], ascending = False)

pvt_borough['Percentage'] = pvt_borough['Percentage'].apply(lambda x: f'{x:.2f}%')
pvt_borough.index = pvt_borough.index.to_series().replace({nan_values: np.nan})

pvt_borough.tail(50)

ID Percentage
Borough                         
Brooklyn        15823     29.44%
Bronx           12268     22.83%
Queens          12169     22.64%
Manhattan       10381     19.32%
Staten Island    2763      5.14%
NaN               260      0.48%
Outside of NYC     77      0.14%

There were 77 incidents <code>Outside of NYC</code> and 260 incidents with missing data. Since (a) the numbers are relatively small (0,52% in total) and (b) we will need borough-specific information for our further analysis, rows with these values will be dropped before we proceed.

In [109]:
df = df[df['Borough'].notna() & ~df['Borough'].isin(['Outside of NYC', 'Unknown'])]
df['Borough'].value_counts()

Brooklyn         15823
Bronx            12268
Queens           12169
Manhattan        10381
Staten Island     2763
Name: Borough, dtype: int64

This looks much better! Let’s continue with <code>Fire battalion</code>, <code>Community district</code>, and <code>Police precinct</code>.
<br><br><code>Fire battalion</code>:

In [110]:
nan_values = 'Unknown'
df['Fire battalion'].fillna(nan_values, inplace = True)

pvt_firebat = pd.pivot_table(
    df,
    values = 'ID',
    index = 'Fire battalion',
    aggfunc = 'count',
)

pvt_firebat['Percentage'] = (pvt_firebat['ID'] / pvt_firebat['ID'].sum()) * 100
pvt_firebat = pvt_firebat.sort_values(by = ['Percentage'], ascending = False)

pvt_firebat['Percentage'] = pvt_firebat['Percentage'].apply(lambda x: f'{x:.2f}%')
pvt_firebat.index = pvt_firebat.index.to_series().replace({nan_values: np.nan})

pvt_firebat

ID Percentage
Fire battalion                 
NaN             7431     13.91%
50              1667      3.12%
54              1500      2.81%
15              1432      2.68%
46              1410      2.64%
18              1356      2.54%
19              1349      2.53%
44              1305      2.44%
38              1226      2.30%
14              1184      2.22%
39              1180      2.21%
3               1167      2.19%
49              1160      2.17%
41              1144      2.14%
17              1112      2.08%
51              1084      2.03%
58              1082      2.03%
26              1070      2.00%
27              1068      2.00%
52              1030      1.93%
12              1016      1.90%
13               969      1.81%
20               961      1.80%
16               959      1.80%
11               942      1.76%
57               923      1.73%
33               907      1.70%
43               904      1.69%
21               900      1.69%
22               846      1.58%
4                830      1.55%
45               825      1.54%
47               794      1.49%
31               791      1.48%
35               780      1.46%
10               778      1.46%
7                754      1.41%
40               752      1.41%
9                694      1.30%
37               661      1.24%
53               654      1.22%
23               647      1.21%
48               624      1.17%
42               604      1.13%
8                600      1.12%
6                555      1.04%
28               554      1.04%
32               481      0.90%
1                399      0.75%
2                327      0.61%
BC                 7      0.01%
0                  3      0.01%
55                 1      0.00%
CD                 1      0.00%
MR                 1      0.00%
OP                 1      0.00%
SC                 1      0.00%
TR                 1      0.00%

<code>Community district</code>:

In [111]:
nan_values = 'Unknown'
df['Community district'].fillna(nan_values, inplace = True)

pvt_comdist = pd.pivot_table(
    df,
    values = 'ID',
    index = 'Community district',
    aggfunc = 'count',
)

pvt_comdist['Percentage'] = (pvt_comdist['ID'] / pvt_comdist['ID'].sum()) * 100
pvt_comdist = pvt_comdist.sort_values(by = ['Percentage'], ascending = False)

pvt_comdist['Percentage'] = pvt_comdist['Percentage'].apply(lambda x: f'{x:.2f}%')
pvt_comdist.index = pvt_comdist.index.to_series().replace({nan_values: np.nan})

pvt_comdist

ID Percentage
Community district                 
1                   5700     10.67%
12                  5129      9.60%
3                   4861      9.10%
5                   4640      8.69%
9                   3505      6.56%
2                   3435      6.43%
4                   3393      6.35%
7                   3228      6.04%
10                  3166      5.93%
11                  2885      5.40%
6                   2765      5.18%
8                   2710      5.07%
13                  1856      3.48%
14                  1710      3.20%
17                  1113      2.08%
16                  1100      2.06%
18                  1069      2.00%
15                   717      1.34%
Out of NYC           215      0.40%
NaN                  207      0.39%

<code>Police precinct</code>:

In [112]:
nan_values = 'Unknown'
df['Police precinct'].fillna(nan_values, inplace = True)

pvt_precinct = pd.pivot_table(
    df,
    values = 'ID',
    index = 'Police precinct',
    aggfunc = 'count',
)

pvt_precinct['Percentage'] = (pvt_precinct['ID'] / pvt_precinct['ID'].sum()) * 100
pvt_precinct = pvt_precinct.sort_values(by = ['Percentage'], ascending = False)

pvt_precinct['Percentage'] = pvt_precinct['Percentage'].apply(lambda x: f'{x:.2f}%')
pvt_precinct.index = pvt_precinct.index.to_series().replace({nan_values: np.nan})

pvt_precinct.tail(25)

ID Percentage
Police precinct                
78               357      0.67%
30               356      0.67%
17               355      0.66%
26               355      0.66%
100              351      0.66%
76               348      0.65%
20               345      0.65%
6                317      0.59%
NaN              200      0.37%
5th               15      0.03%
2                 13      0.02%
0                  9      0.02%
PAPD               5      0.01%
22                 4      0.01%
11                 4      0.01%
7th                3      0.01%
HY2                3      0.01%
12                 2      0.00%
MTS                2      0.00%
PSA7               2      0.00%
21                 1      0.00%
HWY2               1      0.00%
NJ                 1      0.00%
PSA2               1      0.00%
mtn                1      0.00%

After performing the check we can conclude this:
- <b>Fire battalion:</b> As we’ve already seen, there’re many rows without a battalion. I will replace them with a <code>999</code>. Battalions that have an occurrence of less than 327 look suspicious. Probably they are outliers or errors. I will replace them with <code>998</code>.
- <b>Community district:</b> Missing values will be replaced with <code>999</code>. As for <code>Out of NYC</code> I will replace them with <code>998</code>.
- <b>Police precinct:</b> Missing values will be replaced with <code>999</code>. <code>5th</code> and <code>7th</code> are clearly typos. This should be corrected by removing <code>th</code>. Similar to <code>Fire battalion</code> precinct’s that have an occurrence of less than 317 look suspicious. I will replace them with <code>998</code>.

<table>
    <thead>
        <tr><th>Column</th><th>This value</th><th>Will be changed to</th></tr>
    </thead>
    <tbody>
        <tr><td>Fire battalion</td><td>NaN</td><td>999</td></tr>
        <tr><td>Fire battalion</td><td><i>Occurrence of less than 327</i></td><td>998</td></tr>
        <tr><td>Community district</td><td>NaN</td><td>999</td></tr>
        <tr><td>Community district</td><td>Out of NYC</td><td>998</td></tr>
        <tr><td>Police precinct</td><td>5th</td><td>5</td></tr>
        <tr><td>Police precinct</td><td>7th</td><td>7</td></tr>
        <tr><td>Police precinct</td><td>NaN</td><td>999</td></tr>
        <tr><td>Police precinct</td><td><i>Occurrence of less than 317</i></td><td>998</td></tr>
    </tbody>
</table>

## __First transformations: numeric columns cleaned__

In [113]:
# Fire battalion
# Replace 'NaN' and 'Unknown' with '999'
df['Fire battalion'] = df['Fire battalion'].fillna('999').replace('Unknown', '999')

# Replace values with occurrence of less than 327 with '998'
replace_firebat = df['Fire battalion'].value_counts()
replace_firebat = replace_firebat[replace_firebat < 327].index
df['Fire battalion'] = df['Fire battalion'].replace(replace_firebat, '998')

# Community district
# Replace 'NaN' and 'Unknown' with '999' and 'Out of NYC' with '998'
df['Community district'] = df['Community district'].fillna('999').replace({'Unknown': '999', 'Out of NYC': '998'})

# Police precinct
# Remove 'th' from '5th' and '7th'
df['Police precinct'] = df['Police precinct'].str.replace('th', '')

# Replace 'NaN' and 'Unknown' with '999'
df['Police precinct'] = df['Police precinct'].fillna('999').replace('Unknown', '999')

# Replace values with occurrence of less than 200 with '998'
replace_precinct = df['Police precinct'].value_counts()
replace_precinct = replace_precinct[replace_precinct < 200].index
df['Police precinct'] = df['Police precinct'].replace(replace_precinct, '998')

In [114]:
text_check(df, ['Fire battalion', 'Community district', 'Police precinct'])

NON-NUMERIC VALUES

FIRE BATTALION
Numeric values only

COMMUNITY DISTRICT
Numeric values only

POLICE PRECINCT
Numeric values only



Great! We’re done with <code>Borough</code>, <code>Fire battalion</code>, <code>Community district</code>, <code>Police precinct</code>. Before we move on to <code>Incident category</code>, <code>Fire cause</code>, and <code>Fire cause category</code> let’s perform these transformations:
1. Drop the <code>ID</code> column;
2. Assign the correct data type to <code>Date</code>, <code>Fire battalion</code>, <code>Community district</code>, <code>Police precinct</code>.

## __Column <code>ID</code> dropped__

In [115]:
df.drop(['ID'], axis = 1, inplace = True)
df.head()

Year                    Date        Borough Fire battalion  \
0  2016  01/01/2016 12:08:00 AM         Queens             50   
1  2016  01/01/2016 12:13:00 AM  Staten Island             21   
2  2016  01/01/2016 12:21:00 AM         Queens             37   
3  2016  01/01/2016 12:29:00 AM          Bronx             15   
4  2016  01/01/2016 12:32:00 AM         Queens             50   

  Community district Police precinct           Incident category  \
0                  8             107    MD Residential Structure   
1                  2             122    PD Residential Structure   
2                  5             104    MD Residential Structure   
3                 12              47  Burn Notification (Albany)   
4                 12             103    PD Residential Structure   

                                         Fire cause Fire cause category  
0                         Smoking (Cigarette/Cigar)             Smoking  
1                 Incendiary - Combustible Material          Incendiary  
2                 Incendiary - Combustible Material          Incendiary  
3  Albany Burn Notif. Unrelated to a Fire/Explosion               Other  
4                 Incendiary - Combustible Material          Incendiary

## __<i>datatypes</i> corrected__

In [116]:
df.dtypes

Year                    int64
Date                   object
Borough                object
Fire battalion         object
Community district     object
Police precinct        object
Incident category      object
Fire cause             object
Fire cause category    object
dtype: object

In [117]:
df[['Fire battalion', 'Community district', 'Police precinct']] = df[['Fire battalion', 'Community district', 'Police precinct']].astype('int64')

In [118]:
df.dtypes

Year                    int64
Date                   object
Borough                object
Fire battalion          int64
Community district      int64
Police precinct         int64
Incident category      object
Fire cause             object
Fire cause category    object
dtype: object

In [119]:
df['Date'] = pd.to_datetime(df['Date'])

In [120]:
df.head()

Year                Date        Borough  Fire battalion  \
0  2016 2016-01-01 00:08:00         Queens              50   
1  2016 2016-01-01 00:13:00  Staten Island              21   
2  2016 2016-01-01 00:21:00         Queens              37   
3  2016 2016-01-01 00:29:00          Bronx              15   
4  2016 2016-01-01 00:32:00         Queens              50   

   Community district  Police precinct           Incident category  \
0                   8              107    MD Residential Structure   
1                   2              122    PD Residential Structure   
2                   5              104    MD Residential Structure   
3                  12               47  Burn Notification (Albany)   
4                  12              103    PD Residential Structure   

                                         Fire cause Fire cause category  
0                         Smoking (Cigarette/Cigar)             Smoking  
1                 Incendiary - Combustible Material          Incendiary  
2                 Incendiary - Combustible Material          Incendiary  
3  Albany Burn Notif. Unrelated to a Fire/Explosion               Other  
4                 Incendiary - Combustible Material          Incendiary

In [121]:
df.dtypes

Year                            int64
Date                   datetime64[ns]
Borough                        object
Fire battalion                  int64
Community district              int64
Police precinct                 int64
Incident category              object
Fire cause                     object
Fire cause category            object
dtype: object

Looks good! We are not completely done with the <code>Date</code> column, but for the time being let’s leave it as it is and move on to the remaining columns: <code>Incident category</code>, <code>Fire cause</code>, and <code>Fire cause category</code>.

## __Cleaning textual columns__

In [122]:
uvalues_check(df, ['Incident category', 'Fire cause', 'Fire cause category'])

UNIQUE VALUES

INCIDENT CATEGORY
['Aircraft', 'Boat', 'Brush', 'Burn Notification', 'Burn Notification (Albany)', 'Commercial Structure', 'Construction Trailer', 'Construction Vehicle', 'Garage/Shed', 'Government Structure', 'Group Home', 'Hospital', 'Hotel', 'Manhole', 'MD Residential Structure', 'MFA', 'Miscellaneous', nan, 'No Fire', 'Nursing Home', 'Other', 'Outside Rubbish', 'PD Residential Structure', 'Possible Single Room Occupancy', 'Public Hallway', 'Public Structure', 'Religious Structure', 'School', 'Sidewalk', 'Street', 'Threat to Burn', 'Transformer', 'Transit', 'Vehicle'] 

FIRE CAUSE
['Aerosol Can', 'Albany Burn Notif. / Pending Investigation', 'Albany Burn Notif. Related to a Previous BFI Inv.', 'Albany Burn Notif. Unrelated to a Fire/Explosion', 'Appliance Cord', 'Arc Welding', 'Ballast', 'Barbecue - Propane', 'Barbecue - Wood/Charcoal', 'Brands', 'Cable/Satellite Equipment', 'Candle', 'Carbon Monoxide Incident', 'Child Playing w/ Other Open Flame', 'Child Playing w/Ma

In [123]:
dataset_check(
    df, 
    shape_check = False,
    dup_check = False,
    dtypes_check = False,
    missing_values_check = True,
    unique_values_check = False,
    stats_check = False,
    corr_check = False
)

MISSING VALUES:
Incident category      789
Fire cause              68
Fire cause category     56
dtype: int64

MISSING VALUES IN %:
Incident category      1.48%
Fire cause             0.13%
Fire cause category    0.10%
dtype: object



## __Cleaning the <code>Incident category</code> column__

We need to do something with the missing values. There are 789 <code>NaN</code> values in <code>Incident category</code>. Let’s see what are the values in the <code>Fire cause</code> when <code>Incident category</code> is <code>NaN</code>:

In [124]:
df[df['Incident category'].isna()]['Fire cause'].value_counts()

Preliminary Investigation Only    785
Name: Fire cause, dtype: int64

Most of the values are <code>Preliminary Investigation Only</code>. What about the other 4? Looks like they are also <code>NaN</code> in <code>Fire cause</code>.

In [125]:
df[df['Incident category'].isna() & df['Fire cause'].isna()]

Year                Date    Borough  Fire battalion  \
14718  2018 2017-11-27 06:48:00      Bronx               3   
15555  2018 2018-01-05 02:37:00      Bronx              14   
28939  2019 2019-12-14 09:35:37     Queens              53   
28942  2019 2019-12-14 13:00:35  Manhattan             999   

       Community district  Police precinct Incident category Fire cause  \
14718                  10               45               NaN        NaN   
15555                   1               40               NaN        NaN   
28939                  11              111               NaN        NaN   
28942                  11               23               NaN        NaN   

      Fire cause category  
14718                 NaN  
15555                 NaN  
28939                 NaN  
28942                 NaN

In [126]:
# Replace these 4 'NaN' with 'Unknown'
df.loc[df['Incident category'].isna() & df['Fire cause'].isna(), 'Incident category'] = 'Unknown'

# Replace the remaining 'NaN' values with 'Under investigation'
df['Incident category'] = df['Incident category'].fillna('Under investigation')

In [127]:
df['Incident category'].value_counts()

MD Residential Structure          20556
Vehicle                            9958
PD Residential Structure           6779
Commercial Structure               3352
Other                              2348
Outside Rubbish                    2247
Burn Notification                  1363
Burn Notification (Albany)         1280
MFA                                 843
Under investigation                 785
No Fire                             750
Garage/Shed                         586
School                              326
Transit                             267
Brush                               242
Religious Structure                 183
Manhole                             175
Government Structure                173
Group Home                          163
Hotel                               160
Hospital                            152
Public Structure                    121
Sidewalk                            112
Street                               84
Public Hallway                       70


In [128]:
dataset_check(
    df, 
    shape_check = False,
    dup_check = False,
    dtypes_check = False,
    missing_values_check = True,
    unique_values_check = False,
    stats_check = False,
    corr_check = False
)

MISSING VALUES:
Fire cause             68
Fire cause category    56
dtype: int64

MISSING VALUES IN %:
Fire cause             0.13%
Fire cause category    0.10%
dtype: object



We’re done with the <code>Incident category</code> column.

## __Cleaning the <code>Fire cause</code> and <code>Fire cause category</code> columns__

Taking a deeper look at <code>Fire cause</code> and <code>Fire cause category</code> columns I notice that the <i>category</i> column has <i>fewer</i> missing values than the descriptive column. Shouldn’t the <code>Fire cause category</code> be based on the <code>Fire cause</code> column? Let’s take a look at these incidents.

In [129]:
df[df['Fire cause'].isna() & ~df['Fire cause category'].isna()].head(20)

Year                Date        Borough  Fire battalion  \
7302   2016 2016-11-30 09:54:00         Queens              52   
8567   2017 2017-01-29 12:07:09      Manhattan               7   
8570   2017 2017-01-29 15:15:29         Queens              50   
8778   2017 2017-02-07 18:00:00       Brooklyn              43   
8848   2017 2017-02-10 23:09:33         Queens             999   
9944   2017 2017-03-28 18:00:00       Brooklyn              42   
10232  2017 2017-04-11 06:31:41  Staten Island              21   
10382  2017 2017-04-17 22:32:00      Manhattan              16   
10389  2017 2017-04-18 04:47:14         Queens              28   
10677  2017 2017-05-02 01:21:00      Manhattan               4   
10742  2017 2017-05-05 08:16:00       Brooklyn              33   
28826  2019 2019-12-08 17:45:35       Brooklyn              42   

       Community district  Police precinct         Incident category  \
7302                    7              109                   Vehicle   
8567                    5               14      Commercial Structure   
8570                    8              107                   No Fire   
8778                   13               60                     Other   
8848                    9              102  MD Residential Structure   
9944                   11               62                     Other   
10232                   1              122                   No Fire   
10382                   9               30  MD Residential Structure   
10389                   5              104      Commercial Structure   
10677                   3                5      Commercial Structure   
10742                  15               61  MD Residential Structure   
28826                  11               62      Commercial Structure   

      Fire cause   Fire cause category  
7302         NaN                Motors  
8567         NaN                 Other  
8570         NaN               No Fire  
8778         NaN                 Other  
8848         NaN  Cooking Carelessness  
9944         NaN                 Other  
10232        NaN               No Fire  
10382        NaN           Hot Objects  
10389        NaN            Appliances  
10677        NaN               Smoking  
10742        NaN            Incendiary  
28826        NaN            Incendiary

It looks like an error — perhaps the <code>Fire cause</code> field was overlooked when the incident form was filled in.
<br><br>
We can fill it in based on the other values in <code>Fire cause</code> that fall under the corresponding <code>Fire cause category</code>. Of course, under the condition that each category has only one fire cause.

In [130]:
missing_fcc_unique = df[df['Fire cause'].isna() & ~df['Fire cause category'].isna()]['Fire cause category'].unique()
print(missing_fcc_unique)

['Motors' 'Other' 'No Fire' 'Cooking Carelessness' 'Hot Objects'
 'Appliances' 'Smoking' 'Incendiary']


In [131]:
unique_causes_list = [
    category for category in missing_fcc_unique
    if len(df[df['Fire cause category'] == category]['Fire cause'].dropna().unique()) == 1
]

unique_causes_list

['Smoking']

In [132]:
df[df['Fire cause category'] == 'Smoking']['Fire cause'].dropna().unique()

array(['Smoking (Cigarette/Cigar)'], dtype=object)

Looks like only 1 category has 1 description. For <code>Smoking</code> it is <code>Smoking (Cigarette/Cigar)</code>. I will replace <code>Smoking</code> rows with the <code>Smoking (Cigarette/Cigar)</code> value while other missing rows in <code>Fire cause</code> will be replaced with <code>Unknown</code>.
<br><br>
The missing values in <code>Fire cause category</code> will also be replaced with <code>Unknown</code>.

In [133]:
# Fire cause
# Replace 'NaN' in 'Fire cause' with 'Smoking' in 'Fire cause category' with 'Smoking (Cigarette/Cigar)'
df.loc[
    df['Fire cause'].isna() & (df['Fire cause category'] == 'Smoking'),
    'Fire cause'
] = 'Smoking (Cigarette/Cigar)'

# Replace the remaining 'NaN' with 'Unknown'
df['Fire cause'] = df['Fire cause'].fillna('Unknown')

# Fire cause category
df['Fire cause category'] = df['Fire cause category'].fillna('Unknown')

In [134]:
dataset_check(
    df, 
    shape_check = False,
    dup_check = False,
    dtypes_check = False,
    missing_values_check = True,
    unique_values_check = False,
    stats_check = False,
    corr_check = False
)

MISSING VALUES:
0

MISSING VALUES IN %:
0%



## __New section__

In [136]:
df.head(20)

Year                Date        Borough  Fire battalion  \
0   2016 2016-01-01 00:08:00         Queens              50   
1   2016 2016-01-01 00:13:00  Staten Island              21   
2   2016 2016-01-01 00:21:00         Queens              37   
3   2016 2016-01-01 00:29:00          Bronx              15   
4   2016 2016-01-01 00:32:00         Queens              50   
5   2016 2016-01-01 00:46:00      Manhattan               9   
6   2016 2016-01-01 01:24:00       Brooklyn              32   
7   2016 2016-01-01 01:46:00         Queens              46   
8   2016 2016-01-01 02:52:00      Manhattan              16   
9   2016 2016-01-01 03:45:00       Brooklyn              28   
10  2016 2016-01-01 05:31:00      Manhattan               9   
11  2016 2016-01-01 08:00:00         Queens              46   
12  2016 2016-01-01 09:52:00      Manhattan               1   
13  2016 2016-01-01 11:50:00          Bronx              15   
14  2016 2016-01-01 13:21:00          Bronx               3   
15  2016 2016-01-01 13:48:00          Bronx              20   
16  2016 2016-01-01 14:27:00       Brooklyn              57   
17  2016 2016-01-01 14:27:00       Brooklyn              57   
18  2016 2016-01-01 16:27:00         Queens              49   
19  2016 2016-01-01 16:38:00         Queens              46   

    Community district  Police precinct           Incident category  \
0                    8              107    MD Residential Structure   
1                    2              122    PD Residential Structure   
2                    5              104    MD Residential Structure   
3                   12               47  Burn Notification (Albany)   
4                   12              103    PD Residential Structure   
5                    4               14    MD Residential Structure   
6                    2               76    PD Residential Structure   
7                    4              110    PD Residential Structure   
8                   10               32    MD Residential Structure   
9                    4               83    MD Residential Structure   
10                   4               18    MD Residential Structure   
11                   4              110                       Other   
12                   2                5    MD Residential Structure   
13                   7               52  Burn Notification (Albany)   
14                   2               41                       Other   
15                  11               49  Burn Notification (Albany)   
16                   3               79           Burn Notification   
17                   3               79    PD Residential Structure   
18                   1              114             Outside Rubbish   
19                   6              112    MD Residential Structure   

                                           Fire cause   Fire cause category  
0                           Smoking (Cigarette/Cigar)               Smoking  
1                   Incendiary - Combustible Material            Incendiary  
2                   Incendiary - Combustible Material            Incendiary  
3    Albany Burn Notif. Unrelated to a Fire/Explosion                 Other  
4                   Incendiary - Combustible Material            Incendiary  
5                                              Candle            Open Flame  
6                                  Other - Open Flame            Open Flame  
7                           Smoking (Cigarette/Cigar)               Smoking  
8                   Incendiary - Combustible Material            Incendiary  
9                           Smoking (Cigarette/Cigar)               Smoking  
10                                             Candle            Open Flame  
11                                    Other - No Fire               No Fire  
12                              Computer Component(s)    Electrical Devices  
13   Albany Burn Notif. Unrelated to a Fire/Explosion                 Other  
14                

In [143]:
uvalues_check(df, ['Incident category', 'Fire cause category'])

UNIQUE VALUES

INCIDENT CATEGORY
['Aircraft', 'Boat', 'Brush', 'Burn Notification', 'Burn Notification (Albany)', 'Commercial Structure', 'Construction Trailer', 'Construction Vehicle', 'Garage/Shed', 'Government Structure', 'Group Home', 'Hospital', 'Hotel', 'Manhole', 'MD Residential Structure', 'MFA', 'Miscellaneous', 'No Fire', 'Nursing Home', 'Other', 'Outside Rubbish', 'PD Residential Structure', 'Possible Single Room Occupancy', 'Public Hallway', 'Public Structure', 'Religious Structure', 'School', 'Sidewalk', 'Street', 'Threat to Burn', 'Transformer', 'Transit', 'Under investigation', 'Unknown', 'Vehicle'] 

FIRE CAUSE CATEGORY
['Appliances', 'Cooking', 'Cooking Carelessness', 'Electrical', 'Electrical Devices', 'Heaters', 'Hot Objects', 'Ignitable', 'Incendiary', 'Motors', 'Natural Sources', 'No Fire', 'Open Flame', 'Other', 'Smoking', 'Unknown'] 



In [149]:
df.loc[df['Fire cause category'] == 'Appliances', 'Fire cause'].value_counts()

Clothes Dryer - Natural Gas    260
Stove  - Natural Gas           200
Other Appliance                192
Clothes Dryer                  169
Refrigeration Compressor        58
Microwave                       48
Clothes Dryer - Electric        35
Stove - Electric                25
Clothes Washer                  14
Coffee Pot                       9
Dishwasher                       9
Stove - Wood                     8
Washer/Dryer Combo               6
Stove - Coal                     2
Stove - Propane                  2
Unknown                          1
Name: Fire cause, dtype: int64

In [149]:
df.loc[df['Fire cause category'] == 'Appliances', 'Fire cause'].value_counts()

Clothes Dryer - Natural Gas    260
Stove  - Natural Gas           200
Other Appliance                192
Clothes Dryer                  169
Refrigeration Compressor        58
Microwave                       48
Clothes Dryer - Electric        35
Stove - Electric                25
Clothes Washer                  14
Coffee Pot                       9
Dishwasher                       9
Stove - Wood                     8
Washer/Dryer Combo               6
Stove - Coal                     2
Stove - Propane                  2
Unknown                          1
Name: Fire cause, dtype: int64

In [150]:
def checkcheck_check(cat_list):
    df.loc[df['Fire cause category'] == 'Appliances', 'Fire cause'].value_counts()

IndentationError: expected an indented block (1616464593.py, line 1)

In [151]:
def count_fire_causes_by_category(df, categories):
    """
    Iterates through each category in 'Fire cause category' and counts occurrences of values in 'Fire cause'.

    Parameters:
    df (pd.DataFrame): The DataFrame containing the data.
    categories (list): List of categories to iterate through.

    Returns:
    None
    """
    for category in categories:
        print(f'Category: {category}')
        # Filter the DataFrame for the current category
        filtered_df = df[df['Fire cause category'] == category]
        
        # Count occurrences of each unique value in 'Fire cause'
        fire_cause_counts = filtered_df['Fire cause'].value_counts()
        
        # Print the results
        if not fire_cause_counts.empty:
            print(fire_cause_counts, '\n')
        else:
            print('No occurrences found.\n')

# Define your categories
categories = [
    'Appliances', 'Cooking', 'Cooking Carelessness', 'Electrical', 'Electrical Devices',
    'Heaters', 'Hot Objects', 'Ignitable', 'Incendiary', 'Motors', 'Natural Sources',
    'No Fire', 'Open Flame', 'Other', 'Smoking', 'Unknown'
]

# Example DataFrame
data = {
    'Fire cause category': ['Appliances', 'Appliances', 'Cooking', 'Electrical', 'Electrical Devices'],
    'Fire cause': ['Clothes Dryer - Natural Gas', 'Stove - Natural Gas', 'Stove - Electric', 'Microwave', 'Coffee Pot']
}
df = pd.DataFrame(data)

# Run the function
count_fire_causes_by_category(df, categories)

Category: Appliances
Clothes Dryer - Natural Gas    1
Stove - Natural Gas            1
Name: Fire cause, dtype: int64 

Category: Cooking
Stove - Electric    1
Name: Fire cause, dtype: int64 

Category: Cooking Carelessness
No occurrences found.

Category: Electrical
Microwave    1
Name: Fire cause, dtype: int64 

Category: Electrical Devices
Coffee Pot    1
Name: Fire cause, dtype: int64 

Category: Heaters
No occurrences found.

Category: Hot Objects
No occurrences found.

Category: Ignitable
No occurrences found.

Category: Incendiary
No occurrences found.

Category: Motors
No occurrences found.

Category: Natural Sources
No occurrences found.

Category: No Fire
No occurrences found.

Category: Open Flame
No occurrences found.

Category: Other
No occurrences found.

Category: Smoking
No occurrences found.

Category: Unknown
No occurrences found.



In [155]:
def get_fire_cause_counts(df, categories):
    """
    Get the value counts for 'Fire cause' based on the given list of 'Fire cause category'.

    Parameters:
    df (pd.DataFrame): The DataFrame containing the data.
    categories (list of str): The list of categories in 'Fire cause category' for which to get value counts of 'Fire cause'.

    Returns:
    dict: A dictionary where keys are categories and values are Series of value counts of 'Fire cause'.
    """
    results = {}
    for category in categories:
        if category in df['Fire cause category'].unique():
            results[category] = df.loc[df['Fire cause category'] == category, 'Fire cause'].value_counts()
        else:
            results[category] = 'Category not found in dataset'
    return results

In [156]:
category = ['Appliances', 'Cooking', 'Cooking Carelessness', 'Electrical', 'Electrical Devices', 'Heaters', 'Hot Objects', 'Ignitable', 'Incendiary', 'Motors', 'Natural Sources', 'No Fire', 'Open Flame', 'Other', 'Smoking', 'Unknown']
get_fire_cause_counts(df,category)

{'Appliances': Clothes Dryer - Natural Gas    1
 Stove - Natural Gas            1
 Name: Fire cause, dtype: int64,
 'Cooking': Stove - Electric    1
 Name: Fire cause, dtype: int64,
 'Cooking Carelessness': 'Category not found in dataset',
 'Electrical': Microwave    1
 Name: Fire cause, dtype: int64,
 'Electrical Devices': Coffee Pot    1
 Name: Fire cause, dtype: int64,
 'Heaters': 'Category not found in dataset',
 'Hot Objects': 'Category not found in dataset',
 'Ignitable': 'Category not found in dataset',
 'Incendiary': 'Category not found in dataset',
 'Motors': 'Category not found in dataset',
 'Natural Sources': 'Category not found in dataset',
 'No Fire': 'Category not found in dataset',
 'Open Flame': 'Category not found in dataset',
 'Other': 'Category not found in dataset',
 'Smoking': 'Category not found in dataset',
 'Unknown': 'Category not found in dataset'}

## Rest

In [135]:
df['Fire battalion'].unique()

array([ 50,  21,  37,  15,   9,  32,  46,  16,  28,   1,   3,  20,  57,
        49,  10,  14,  58,  43,  48,  39,  44,   6,   4,  27,  45,  13,
         8,  47,  31,  51,  54,  42,   7,  26,  19,  52,  35,  12,  23,
        18,  40,  53,  38,  41,  11,  22,  33,  17,   2, 998, 999],
      dtype=int64)

In [136]:
pivot1 = df.pivot_table(index = 'Fire battalion',
               columns = 'Borough',
               values = 'ID',
               aggfunc = 'count',
               fill_value = 0,
               margins = True, 
               margins_name = 'Total')

pivot1.sort_values(by=['Total'], ascending=False)

KeyError: 'ID'

In [ ]:
pivot2 = df.pivot_table(
    index = 'Borough',
    values = 'ID',
    aggfunc = 'count',
    #fill_value = 0,
    margins = True,
    #margins_name = 'Total',
)

pivot2.sort_values(by = 'ID', ascending = False)

In [ ]:
outsideofNYC = df[df['Borough'] == 'Outside of NYC']
outsideofNYC.head(20)

In [ ]:
outsideofNYC[outsideofNYC['Borough'] != 'Outside of NYC']

In [ ]:
outsideofNYC['Fire cause category'].unique()

In [ ]:
outsideofNYC['Fire cause category'].unique()

In [ ]:
outsideofNYC['Fire cause category'].unique()

In [ ]:
outsideofNYC['Fire cause category'].unique()

In [ ]:
outsideofNYC['Fire cause category'].unique()

In [ ]:
77/53481*100

In [ ]:
# Heatmap example
sns.heatmap(pivot1, cmap = 'RdYlGn');

In [ ]:
pivot3 = df.pivot_table(index = 'Police precinct',
               columns = 'Borough',
               values = 'ID',
               aggfunc = 'count',
               fill_value = 0,
               margins = True, 
               margins_name = 'Total')

pivot3b = pivot3.sort_values(by=['Total'], ascending=False)
pivot3b[pivot3b['Total'] <= 300].index

In [ ]:
pivot3b[pivot3b['Total'] <= 400]['Total']#.sum()

In [ ]:
pivot4 = df.pivot_table(index = 'Community district',
               columns = 'Borough',
               values = 'ID',
               aggfunc = 'count',
               fill_value = 0,
               margins = True, 
               margins_name = 'Total')

pivot4.sort_values(by=['Total'], ascending=False)

In [ ]:
df[df['Fire battalion'] == '50']

In [ ]:
df = pd.read_csv(path, parse_dates=['Incident_DateTime'])
df.head()

In [ ]:
cond = df['Incident_DateTime'].dt.dayofweek == 1
df[cond]

In [ ]:
df['Incident_DateTime'].max()

In [ ]:
print(np.sort(df['Case_Year'].unique()))

In [ ]:
df.sort_values(by = ['Case Number', 'Incident_DateTime']).head(20)

In [ ]:
groupby1 = df.groupby(by = ['Case Number']).count()
groupby1['Incident_DateTime'].sort_values(ascending = False)

In [ ]:
df[df['Case Number'] == 20203].sort_values(by = 'Case_Year')